File: front/central/src/shared/utils/rate-pricing.test.ts
import { describe, it, expect } from 'vitest';
import { rateGuideCost, rateCarrierFee, rateTotalCost, rateAppliesCOD, rateBreakdown } from './rate-pricing';

const coordinadora = {
    flete: 7796,
    minimumInsurance: 580,
    extraInsurance: 920,
    cod: true,
    codCarrierFee: 6116,
    codProbabilityMargin: 500,
};

const servientrega = {
    flete: 15390,
    minimumInsurance: 0,
    extraInsurance: 0,
    cod: true,
    codCarrierFee: 8499,
    codProbabilityMargin: 500,
};

const sinCOD = {
    flete: 11424,
    minimumInsurance: 0,
    extraInsurance: 1500,
    cod: false,
};

describe('rate-pricing', () => {
    it('suma flete y seguro minimo siempre', () => {
        expect(rateGuideCost(coordinadora)).toBe(8376);
    });

    it('suma el margen COD solo cuando hay contra entrega', () => {
        expect(rateGuideCost(coordinadora, { cod: true })).toBe(8876);
        expect(rateGuideCost(coordinadora, { cod: false })).toBe(8376);
    });

    it('no aplica margen COD a una tarifa que no soporta contra entrega', () => {
        expect(rateAppliesCOD(sinCOD, { cod: true })).toBe(false);
        expect(rateGuideCost(sinCOD, { cod: true })).toBe(11424);
    });

    it('suma el seguro adicional solo si se asegura', () => {
        expect(rateGuideCost(coordinadora, { insured: true })).toBe(9296);
        expect(rateGuideCost(sinCOD, { insured: true })).toBe(12924);
    });

    it('la comision del carrier solo cuenta con contra entrega', () => {
        expect(rateCarrierFee(coordinadora, { cod: true })).toBe(6116);
        expect(rateCarrierFee(coordinadora, { cod: false })).toBe(0);
    });

    it('el total suma guia mas comision', () => {
        expect(rateTotalCost(coordinadora, { cod: true })).toBe(14992);
        expect(rateTotalCost(servientrega, { cod: true })).toBe(24389);
    });

    it('el desglose suma exactamente el total', () => {
        const d = rateBreakdown(coordinadora, { cod: true, insured: true });
        expect(d.flete).toBe(8296);
        expect(d.flete + d.minimumInsurance + d.extraInsurance).toBe(d.guideCost);
        expect(d.guideCost + d.carrierFee).toBe(d.total);
        expect(d.total).toBe(rateTotalCost(coordinadora, { cod: true, insured: true }));
    });

    it('el desglose sin contra entrega no trae margen ni comision', () => {
        const d = rateBreakdown(coordinadora, { cod: false });
        expect(d.flete).toBe(7796);
        expect(d.carrierFee).toBe(0);
        expect(d.total).toBe(8376);
    });

    it('tolera tarifas incompletas', () => {
        expect(rateGuideCost(null)).toBe(0);
        expect(rateGuideCost({}, { cod: true, insured: true })).toBe(0);
        expect(rateTotalCost({ flete: 1000 } as any, { cod: true })).toBe(1000);
    });
});

File: mobile/mobile_central/test/shared/utils/rate_pricing_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_central/services/modules/shipments/domain/entities.dart';
import 'package:mobile_central/shared/utils/rate_pricing.dart';

EnvioClickRate buildRate({
  double flete = 10000,
  double? minimumInsurance = 6530,
  double? extraInsurance = 4000,
  bool? cod = true,
  double? codCarrierFee = 3000,
  double? codProbabilityMargin = 1500,
}) {
  return EnvioClickRate(
    idRate: 1,
    idProduct: 1,
    product: 'Estandar',
    idCarrier: 1,
    carrier: 'Interrapidisimo',
    flete: flete,
    deliveryDays: 2,
    quotationType: 'normal',
    minimumInsurance: minimumInsurance,
    extraInsurance: extraInsurance,
    cod: cod,
    codCarrierFee: codCarrierFee,
    codProbabilityMargin: codProbabilityMargin,
  );
}

void main() {
  group('RatePricing replica la formula de rate-pricing.ts', () {
    test('sin contra entrega ni seguro extra suma flete y seguro minimo', () {
      final rate = buildRate();
      const options = RatePricingOptions();
      expect(RatePricing.guideCost(rate, options), 16530);
      expect(RatePricing.carrierFee(rate, options), 0);
      expect(RatePricing.totalCost(rate, options), 16530);
    });

    test('el seguro minimo nunca se omite', () {
      final rate = buildRate(minimumInsurance: 6530);
      expect(RatePricing.guideCost(rate, const RatePricingOptions()), 16530);
    });

    test('asegurar suma el seguro extra', () {
      final rate = buildRate();
      const options = RatePricingOptions(insured: true);
      expect(RatePricing.guideCost(rate, options), 20530);
    });

    test('contra entrega suma el margen a la guia y la comision al total', () {
      final rate = buildRate();
      const options = RatePricingOptions(cod: true);
      expect(RatePricing.guideCost(rate, options), 18030);
      expect(RatePricing.carrierFee(rate, options), 3000);
      expect(RatePricing.totalCost(rate, options), 21030);
    });

    test('si la tarifa no soporta cod no se aplica margen ni comision', () {
      final rate = buildRate(cod: false);
      const options = RatePricingOptions(cod: true);
      expect(RatePricing.appliesCod(rate, options), isFalse);
      expect(RatePricing.guideCost(rate, options), 16530);
      expect(RatePricing.carrierFee(rate, options), 0);
    });

    test('el desglose cuadra con el total', () {
      final rate = buildRate();
      const options = RatePricingOptions(cod: true, insured: true);
      final breakdown = RatePricing.breakdown(rate, options);
      expect(breakdown.flete, 11500);
      expect(breakdown.minimumInsurance, 6530);
      expect(breakdown.extraInsurance, 4000);
      expect(breakdown.carrierFee, 3000);
      expect(breakdown.guideCost, 22030);
      expect(breakdown.total, 25030);
      expect(breakdown.total, RatePricing.totalCost(rate, options));
    });

    test('valores nulos se tratan como cero', () {
      final rate = buildRate(minimumInsurance: null, extraInsurance: null, codProbabilityMargin: null, codCarrierFee: null);
      const options = RatePricingOptions(cod: true, insured: true);
      expect(RatePricing.guideCost(rate, options), 10000);
      expect(RatePricing.totalCost(rate, options), 10000);
    });
  });
}
